### PR TITLE
[DTensor] Fix slow sharding prop for stack

### LINKED ISSUE #169445: Quadratic complexity in gradient clipping
### 🐛 Describe the bug

When gradient clipping using FSDP2, we torch.stack a list of O(named parameters) DTensors:
https://github.com/pytorch/pytorch/blob/ef019d1d431c4c5a95b594cb90d40a50cd00f5e4/torch/nn/utils/clip_grad.py#L107 

I tested torch.stack with the `DTensor` type and found that it is quadratic in the number of tensors (with regular `tensor` types this is linear):

Image

For MoE models, the number of named parameters can be very large (e.g. Qwen/Qwen3-30B-A3B with lora has around 20k). By my extrapolation, this means that it will take >90 minutes just to clip the gradients!

### Versions

PyTorch version: 2.9.0+cu128
Is debug build: False
CUDA used to build PyTorch: 12.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.1 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: Could not collect
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.12.11 (main, Aug 28 2025, 17:07:59) [Clang 20.1.4 ] (64-bit runtime)
Python platform: Linux-5.15.0-161-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to:
GPU models and configuration:
GPU 0: NVIDIA H100 80GB HBM3
GPU 1: NVIDIA H100 80GB HBM3

Nvidia driver version: 570.124.06
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.8.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture: x86_64
CPU op-mode(s): 32-bit, 64-bit
Address sizes: 46 bits physical, 57 bits virtual
Byte Order: Little Endian
CPU(s): 104
On-line CPU(s) list: 0-103
Vendor ID: GenuineIntel
Model name: Intel(R) Xeon(R) Platinum 8470
CPU family: 6
Model: 143
Thread(s) per core: 1
Core(s) per socket: 52
Socket(s): 2
Stepping: 8
BogoMIPS: 4000.00
Flags: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 invpcid_single cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities ibpb_exit_to_user
Virtualization: VT-x
L1d cache: 4.9 MiB (104 instances)
L1i cache: 3.3 MiB (104 instances)
L2 cache: 208 MiB (104 instances)
L3 cache: 210 MiB (2 instances)
NUMA node(s): 8
NUMA node0 CPU(s): 0-12
NUMA node1 CPU(s): 13-25
NUMA node2 CPU(s): 26-38
NUMA node3 CPU(s): 39-51
NUMA node4 CPU(s): 52-64
NUMA node5 CPU(s): 65-77
NUMA node6 CPU(s): 78-90
NUMA node7 CPU(s): 91-103
Vulnerability Gather data sampling: Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit: Not affected
Vulnerability L1tf: Not affected
Vulnerability Mds: Not affected
Vulnerability Meltdown: Not affected
Vulnerability Mmio stale data: Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed: Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1: Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2: Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds: Not affected
Vulnerability Tsa: Not affected
Vulnerability Tsx async abort: Not affected
Vulnerability Vmscape: Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] Could not collect
[conda] Could not collect

cc @ezyang @gchanan @kadeng @msaroufim @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @dcci @zhaojuanmao @mrshenli @rohan-varma @chauhang @mori360 @tianyu-l @XilunWu @SherlockNoMad

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #169519

As identified in the original issue, there is quadratic complexity in
the number of input tensors, due to an improperly written sharding prop
rule.

The previous code generated N output strategies for the stack op, one
based on each of the original N input strategies.  However, Each of the
N output strategies was the same.  The heuristic in the stack rule is to
find one of the N inputs and follow that one.

We now just generate one output strategy.

Fixes #169445

### CODE REVIEW

File: torch/distributed/tensor/_ops/_tensor_ops.py [hunk 1]
```diff
@@ -759,9 +759,11 @@ def stack_strategy(op_schema: OpSchema) -> StrategyType:
     input_tuple_strategy = args_schema[0]
     if not isinstance(input_tuple_strategy, TupleStrategy):
         raise AssertionError(f"Expected TupleStrategy, got {input_tuple_strategy}")
-    first_input_strategy = input_tuple_strategy.children[0]
-    if not isinstance(first_input_strategy, OpStrategy):
-        raise AssertionError(f"Expected OpStrategy, got {first_input_strategy}")
+    input_strategies: list[OpStrategy] = []
```
Comment (wconstab) [inline torch/distributed/tensor/_ops/_tensor_ops.py:762]: this part was just to make mypy happy below since children are listed as 'StrategyType' which can be Tuple Strategy or OpStrategy,  need to ensure they are all  OpStrategy...

**Review by zpcore — APPROVED** [review]
LGTM!

File: torch/distributed/tensor/_ops/_tensor_ops.py [hunk 2]
```diff
@@ -784,22 +786,18 @@ def stack_strategy(op_schema: OpSchema) -> StrategyType:
...
     # be normalized with the new Shard placement
     follow_placements = shift_shard_dims_after_insert(follow_placements, dim)
-
-    for strategy in input_tuple_strategy.children:
-        if not isinstance(strategy, OpStrategy):
-            raise AssertionError(f"Expected OpStrategy, got {type(strategy)}")
-        output_spec = DTensorSpec(mesh, tuple(follow_placements))
-        redistribute_cost = []
-        for input_spec in input_specs:
-            cost = generate_redistribute_costs(strategy, input_spec)
```
Comment (wconstab) [inline torch/distributed/tensor/_ops/_tensor_ops.py:794]: @zpcore one thing i would like to confirm is, this old code looks incorrect to me, in addition to being slower.

we should never be generating the redistribute cost from input 2's placement to input1's dst spec, right?  so using 'strategy' here was a bug?

Comment (zpcore) [reply torch/distributed/tensor/_ops/_tensor_ops.py:794]: Good catch! Should be:
```
for idx, input_spec in enumerate(input_specs):
            cost = generate_redistribute_costs(input_tuple_strategy.children[idx], input_spec)
```

Image

**Review by malfet — APPROVED** [review]

**Review by albanD — APPROVED** [review]
Sounds good!
Any reason this strategy is not shared with cat() ?
